### PR TITLE
scripts/dts: Build a list of flash area device names

### PR DESCRIPTION
**Pull request by Laczen**
The current flash processing of the dts does not give information
about the device driver for the different images.

The modification generates the following defines:

`DT_FLASH_AREA_IMAGE_0_DEV="NRF_FLASH_DRV_NAME"`
`DT_FLASH_AREA_IMAGE_1_DEV="NRF_FLASH_DRV_NAME"`
`DT_FLASH_AREA_IMAGE_SCRATCH_DEV="NRF_FLASH_DRV_NAME"`
`DT_FLASH_AREA_MCUBOOT_DEV="NRF_FLASH_DRV_NAME"`
`DT_FLASH_AREA_STORAGE_DEV="NRF_FLASH_DRV_NAME"`

This allows to open the correct device when trying to write to
a specific flash area.

### CODE REVIEW

File: scripts/dts/extract/flash.py [hunk 1]
```diff
@@ -91,6 +91,36 @@ def _extract_partition(self, node_address):
 
         insert_defs("DT_FLASH_AREA", prop_def, prop_alias)
 
+    def _extract_partition_dev(self, node_address):
+        prop_def = {}
+        prop_alias = {}
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: Could get rid of these variables and just do

```python
insert_defs("DT_FLASH_AREA_DEV", {label: name}, {})`
```

File: scripts/dts/extract/flash.py [hunk 2]
```diff
@@ -91,6 +91,36 @@ def _extract_partition(self, node_address):
...
+
+        # For now assume node_address is something like:
+        # /flash-controller@4001E000/flash@0/partitions/partition@fc000
+        # first we go up 2 levels to get the flash, check its compat
+        #
+        # The flash controller might be the flash itself (for cases like NOR
+        # flashes), for the case of 'soc-nv-flash' we assume its the parent
+        # of the flash node.
+        ctrl_addr = '/' + '/'.join(node_address.split('/')[1:-2])
+        if get_compat(ctrl_addr) == "soc-nv-flash":
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: `'` quotes would be better. Consistency is handy for `grep`ing.

File: scripts/dts/extract/flash.py [hunk 3]
```diff
@@ -91,6 +91,36 @@ def _extract_partition(self, node_address):
...
+            ctrl_addr = '/' + '/'.join(node_address.split('/')[1:-3])
+
+        node = reduced[ctrl_addr]
+        name = "\"{}\"".format(node['props']['label'])
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: Could make it `'"{}"'`.

File: scripts/dts/extract/flash.py [hunk 4]
```diff
@@ -91,6 +91,36 @@ def _extract_partition(self, node_path):
...
+
+        label = self.get_label_string(["DT_FLASH_AREA"] + [partition_name] + ["DEV",])
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: Can be simplified to

```python
label = self.get_label_string(["DT_FLASH_AREA", partition_name, "DEV"])
```

No need to do `[foo,]` for a single-element list, though you see it in some other places. Just `[foo]` works fine.

File: scripts/dts/extract/flash.py [hunk 5]
```diff
@@ -91,6 +91,36 @@ def _extract_partition(self, node_path):
...
+        prop_def[label] = name
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: (Could get rid of this along with getting rid of the `prop_def` variable.)

File: scripts/dts/extract/flash.py [hunk 6]
```diff
@@ -99,7 +129,12 @@ def extract_partition(self, node_path):
         prop_alias = {}
         node = reduced[node_path]
 
+<<<<<<< HEAD
         self._extract_partition(node_path)
+=======
+        self._extract_partition(node_address)
+        self._extract_partition_dev(node_address)
+>>>>>>> scritps/dts: Build a list of flash area device names
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: Accidentally committed merge conflict. :)

**Review by ulfalizer — CHANGES_REQUESTED** [review]
Some code nits.

File: scripts/dts/extract/flash.py (hunk 6 above)
Comment (Laczen) [reply scripts/dts/extract/flash.py]: @ulfalizer, thanks a lot for the review. I have to say, I don't really understand how these scripts are working. I just copied the _extract_partition and removed everything I didn't need.

Comment (ulfalizer) [reply scripts/dts/extract/flash.py]: @Laczen 
Yeah, no worries. Many people who haven't done that much Python have been working on these, so there's some weirdness in there. :)

Comment (ulfalizer) [reply scripts/dts/extract/flash.py]: I'm trying to make them a bit easier to read. Many parts are super hard for me to understand too.

**Review by dbkinder — COMMENTED** [review]
No doc changes for me to review :)

**Comment by Laczen** [comment]
@dbkinder, indeed, you were requested a review because of a commit that was not needed in this PR, my apologies for this.

File: scripts/dts/extract/flash.py [hunk 7]
```diff
@@ -91,6 +91,27 @@ def _extract_partition(self, node_path):
 
         insert_defs("DT_FLASH_AREA", prop_def, prop_alias)
 
+    def _extract_partition_dev(self, node_address):
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: Missed renaming: `node_address` -> `node_path`

File: scripts/dts/extract/flash.py [hunk 8]
```diff
@@ -91,6 +91,27 @@ def _extract_partition(self, node_path):
...
+        # For now assume node_address is something like:
+        # /flash-controller@4001E000/flash@0/partitions/partition@fc000
+        # first we go up 2 levels to get the flash, check its compat
+        #
+        # The flash controller might be the flash itself (for cases like NOR
+        # flashes), for the case of 'soc-nv-flash' we assume its the parent
+        # of the flash node.
+        ctrl_addr = '/' + '/'.join(node_address.split('/')[1:-2])
+        if get_compat(ctrl_addr) == 'soc-nv-flash':
+            ctrl_addr = '/' + '/'.join(node_address.split('/')[1:-3])
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: Could do this maybe. I didn't catch right away that the calculated value was overridden.

```python
if get_compat(ctrl_addr) == 'soc_nv_flash':
    ...
else:
    ...
```

Nice comment above by the way!

**Review by ulfalizer — CHANGES_REQUESTED** [review]

**Comment by ulfalizer** [comment]
Heh, scrap that overriden comment. Code-reading failure. :)

File: scripts/dts/extract/flash.py [hunk 9]
```diff
@@ -91,6 +91,27 @@ def _extract_partition(self, node_path):
...
+        # For now assume node_path is something like:
+        # /flash-controller@4001E000/flash@0/partitions/partition@fc000
+        # first we go up 2 levels to get the flash, check its compat
+        #
+        # The flash controller might be the flash itself (for cases like NOR
+        # flashes), for the case of 'soc-nv-flash' we assume its the parent
+        # of the flash node.
+        ctrl_addr = '/' + '/'.join(node_path.split('/')[1:-2])
+        if get_compat(ctrl_addr) == 'soc-nv-flash':
+            ctrl_addr = '/' + '/'.join(node_path.split('/')[1:-3])
```
Comment (ulfalizer) [inline scripts/dts/extract/flash.py]: Could do `ctrl_addr = get_parent_path(ctrl_path)` here too (and do `get_parent_path(get_parent_path(node_path))` above).

**Review by ulfalizer — CHANGES_REQUESTED** [review]
Just one nit. Sorry for not spotting it earlier

**Comment by carlescufi** [comment]
@Laczen can you address the outstanding comments?

**Comment by galak** [comment]
How or where is the use of this set of defines?

**Comment by Laczen** [comment]
@ulfalizer, could you please explain the code change you wanted?
@galak, the availability of the defines allow you to get the correct device for each of the defined areas in flash. As an example suppose you want to use the storage partition for a nvs file_system:
``` nvs_int(&fs, DT_FLASH_AREA_STORAGE_DEV)```

**Comment by Laczen** [comment]
@galak, the defines also allow you to generate a flashmap with area_id linked to a known flash area without the need of a script to generate the flash_map.

**Comment by Laczen** [comment]
@ulfalizer, I know there might be other ways to achieve the same thing, but please don't let this be a stopper.
@galak, The use of this is plenty: if you would like to use whatever area to write something to you can just get the device using: ```device_get_binding(DT_FLASH_AREA_XXX_DEV)```, no matter where the ```FLASH_AREA``` is located (on chip flash, spi flash, ...). This is a lot simpler and clearer than using the area_id. Using this together with the flash page layout allows you to get page information directly.
@nvlsianpu, using these defines it is possible to rework the flashmap and get a fixed link again between area_id and ```DT_FLASH_AREA_XXX```.

**Comment by nvlsianpu** [comment]
I'm not convinced to using this. But see no serious obstacles to have in codebase - as discuses with @Laczen - there is need for it in order to build low flash footprint bootloader

As the Nordic I had introduced flash-map structure as common place for NVM boundaries and properties.
I prefer to use it and flash_area API together as abstraction above flash drivers. But I also see the point of extracting flash device data (see PR: https://github.com/zephyrproject-rtos/zephyr/pull/13303).
flash_map has potential to be inter-application flash description (for multi application sw arch on the sam e SoC). It is already also required for mcuboot. I is also salable for multi mage mcuboot extension.

**Comment by Laczen** [comment]
@nvlsianpu, thanks for the comment. The idea is not to replace flash_map, it only adds a dev to each of the already defined DT_FLASH_AREA_XXX definitions. This could be used directly (e.g. for nvs) or as a base to generate the default_flash_map.

**Comment by ulfalizer** [comment]
> @ulfalizer, I know there might be other ways to achieve the same thing, but please don't let this be a stopper.

Just asking for `get_parent_path()`, instead of re-implementing it.

**Review by ulfalizer — APPROVED** [review]
`get_parent_path()` nit fixed. Other people know more about how necessary the generated `#define`s are.

**Comment by Laczen** [comment]
@nashif, @carlescufi, @galak, I need this to improve the nvs sample and also to use nvs as a settings backend.

**Comment by nvlsianpu** [comment]
Right now you can use what was introduced here https://github.com/zephyrproject-rtos/zephyr/pull/13303. 

I know that it is not exactly your plan - but I want to avoid cut of future possibility (because central source of data is easy to be replaced in product lifetme, if flash_map become a inter-app shared resource). Maybe using of flash_map or your macros (mean what this PR introduce) can be warped by macros so API will looks the same but user will chose whether want to use centralized source of data or distributed.

**Comment by nvlsianpu** [comment]
```
#ifdef CONFIG_xxx
#define FLASH_AREA_GET_DEVICE(area_name) flash_area_get_device(area_name)
#else
#define	FLASH_AREA_GET_DEVICE(area_name) /* something which uses new macros */
#endif
```

**Comment by Laczen** [comment]
> ```
> #ifdef CONFIG_xxx
> #define FLASH_AREA_GET_DEVICE(area_name) flash_area_get_device(area_name)
> #else
> #define	FLASH_AREA_GET_DEVICE(area_name) /* something which uses new macros */
> #endif
> ```

Yeah sure, and by default enable flash_map in all config files?

**Comment by nvlsianpu** [comment]
^^ Yes

**Comment by Laczen** [comment]
What I propose is both temporary and final:
a. As long as there is no real centralized flashmap the users can decide to use flashmap or the direct macro supplied device names.
b. As soon as a real centralized flashmap (could be something like the partition table on ESP) is available the direct macro supplied device name is used to get the device where the flashmap is located (this will always be required to be generated). The users would then only use the flashmap interface, and all information regarding the flash partitioning can be removed from the dts.

**Comment by baw-serafin** [comment]
Just wanted to add that this seems very reasonable and useful for people like us who use external flashes as well as internal one. The flahmap API seems a ittle overkill for a lot of cases were you only want to get the information from the dt. But I have not yet looked to deep into that one.

**Comment by nvlsianpu** [comment]
PR In current shape makes possible to introduce distributed incompatibility in the code with ["Inter-applications flash layout exchange mechanism" feature request](https://github.com/zephyrproject-rtos/zephyr/issues/6152). 

It is still possible to have direct references (introduced by this PR) to driver names as optional (with I suggested a few comment s above).

@baw-serafin You can use `struct device *flash_area_get_device(const struct flash_area *fa);` API.

**Comment by baw-serafin** [comment]
@nvlsianpu Can't find any documentation to either `flash_area_get_device` or `struct flash_area`

The flash_map is also mostly undocumented. @Laczen approach just feels much more zephyr like. The other APIs just show that they were designed for something else.

But well as long as I can get the device without to much hassle I don't mind.

**Comment by Laczen** [comment]
> PR In current shape makes possible to introduce distributed incompatibility in the code with ["Inter-applications flash layout exchange mechanism" feature request](https://github.com/zephyrproject-rtos/zephyr/issues/6152).
> 
> It is still possible to have direct references (introduced by this PR) to driver names as optional (with I suggested a few comment s above).
> 
> @baw-serafin You can use `struct device *flash_area_get_device(const struct flash_area *fa);` API.


@nvlsianpu, what you are saying is wrong, correct would be:

**This PR allows the "Inter-application flash layout exchange mechanism" to be located on whatever flash device you would like. It also provides easy access to any dts defined partition that needs to be outside the "Inter-application flash layout exchange mechanism" (because it cannot be accessed by the bootloader). As long as the "Inter-application flash layout exchange mechanism" is not in place it allows easy access to all dts defined partitions.**

**Pull request closed without merging**